src/wallet/wallet.cpp

```diff
@@ -947,7 +963,10 @@ bool CWallet::AddToWalletIfInvolvingMe(const CTransactionRef& ptx, const CBlockI
             if (pIndex != nullptr)
                 wtx.SetMerkleBranch(pIndex, posInBlock);
 
-            return AddToWallet(wtx, false);
+            if (IsMine(tx) || IsFromMe(tx))
+                return AddToWallet(wtx, false);
+
+            return RemoveFromWallet(tx);
```

Comment:
This looks a bit invasive.
I haven't looked closer at it, but would it not be possible to remove transaction with the removed addresses during `LoadToWallet()` rather then an expansive rescan?

Reply:
Without this, it will remove the balance and addresses from the wallets but it will leave all the transactions with their ids in the wallet so with this function it will remove them from the wallet during the rescan.  I am not sure if there is a faster way to do it.

Reply:
I don't see the reason why a rescan would be required. IMO it's only required after adding a script where its creation-timestamp lies in the past (expected historical transactions). You don't need to go through the whole (or part of) the blockchain history just to know that a transaction in your wallet doesn't belong to you anymore. Also, don't forget that rescans are incompatible for pruned peers.

It should be sufficient to loop though `mapWallet` and remove transactions that are no longer `IsMine(tx) || IsFromMe(tx)` (after the deletion of one or many scripts/addresses).